torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -1939,6 +1887,241 @@ std::pair<std::vector<kir::Val*>, bool> Index::getConsumerRootPredIndices(
+                        ->as<kir::IterDomain>();
+    ref_id_to_ind_map[ref_axis] = loop_to_ind_map[loops[loop_i]];
+  }
+
+  auto consumer_tv = kir_consumer_tv->fuserTv();
+
+  // Map reference tensor to consumer
+  std::unordered_map<IterDomain*, IterDomain*> root_ref_to_consumer;
+  for (auto c_root : consumer_tv->getMaybeRFactorDomain()) {
+    auto concrete_id = gpu_lower->caIndexMap().getConcreteMappedID(c_root);
+    auto ref_id_it = reference_id_map.find(concrete_id);
+    if (ref_id_it != reference_id_map.end()) {
+      root_ref_to_consumer[ref_id_it->second] = c_root;
+    }
+  }
+
+  BestEffortReplay replay_consumer_as_ref(
+      consumer_tv->domain()->domain(),
+      reference_domain->domain(),
+      root_ref_to_consumer);
+
+  const auto& ref_2_consumer = replay_consumer_as_ref.getReplay();
+
+  // Generate halo information for reference.
+  updateHaloInfoForReference(reference, consumer_tv);
+
+  std::unordered_map<kir::IterDomain*, kir::Val*> reference_halo_extent_map;
+
+  const auto& halo_info = gpu_lower->haloInfo();
+
+  // Generate map from reference iter domains to halo extents
+  for (auto entry : ref_2_consumer) {
+    auto ref_id = entry.first;
+    auto extent = halo_info.getExtent(ref_id);
+    if (extent != nullptr) {
+      reference_halo_extent_map[gpu_lower->lowerValue(ref_id)
+                                    ->as<kir::IterDomain>()] =
+          gpu_lower->lowerValue(extent);
+    }
+  }
+
+  // Index into the reference tensor
+  auto ref_indexing = getReferenceIndexing(
+      loops,
+      reference_domain,
+      ref_id_to_ind_map,
+      {},
+      reference_halo_extent_map);
+
+  // If we are initializing a reduction buffer and the tensor has a
+  // rfactor root, the predicate should be based on the rfactor root.
+  const auto root_domain = reference_domain->getRootDomain();
+
+  auto contig_id_infos = getPredicateContigIds(reference_domain->domain());
+
+  // Roots in contiguous processing is based on reference roots, want to convert
+  // these to concrete roots, flip reference's concrete_to_id map
+  decltype(reference_id_map) ref_id_to_concrete;
+  std::transform(
+      reference_id_map.begin(),
+      reference_id_map.end(),
+      std::inserter(ref_id_to_concrete, ref_id_to_concrete.begin()),
+      [](auto entry) { return std::make_pair(entry.second, entry.first); });
+
+  std::vector<std::unordered_set<IterDomain*>> handeled_roots;
+
+  std::vector<kir::Bool*> predicates;
+
+  for (auto contig_id_entry : contig_id_infos) {
+    auto contig_id = contig_id_entry.contig_id;
+    if (contig_id->isBroadcast() ||
+        gpu_lower->trivialReductionInfo().isDerived(contig_id)) {
+      continue;
+    }
+
+    auto root_ids = contig_id_entry.root_ids;
+    auto kir_contig_id =
+        gpu_lower->lowerValue(contig_id)->as<kir::IterDomain>();
+
+    const auto it = ref_indexing.indexMap().find(kir_contig_id);
+    if (it == ref_indexing.indexMap().end() || it->second->isZeroInt()) {
```

Comment:
It would be helpful to have some comments here. For example, (correct me if I'm wrong), `it == ref_indexing.indexMap().end()` means that there's no corresponding ID in the reference due to broadcast, so no predicate is necessary for the ID. 

Similarly, `it->second->isZeroInt()` means that the corresponding ID in the reference tensor is a broadcast axis, so no predicate is necessary in this case too.